### ISSUE TITLE
Could Factory be updated to use Ember's CoreObject, to avoid linting issues introduced in Ember 2.18.0?

### ISSUE THREAD

**Issue by BillyRayPreachersSon** [samselikoff/ember-cli-mirage#1248]
Ember 2.18.0 introduced some new linting rules, which gives pretty good guidance about certain coding errors.

One of them is where objects could potentially [leak state](https://github.com/ember-cli/eslint-plugin-ember/blob/master/docs/rules/avoid-leaking-state-in-ember-objects.md) by exposing arrays or objects as default properties.

The fix for this issue is to move array properties into the `init()` method of the object, so this:

```
export default Foo.extend({
    items: ['some', 'things']
});
```

becomes this:

```
export default Foo.extend({
    init() {
        this._super(...arguments);
        this.items = ['some', 'things'];
    }
});
```

This new rule is only run on objects it deems to be Ember objects, but the detection of this seems a bit flaky: it's detecting `Factory` as one of these when it is not, and thus build process breaks with an error that is not really relevant.

I [raised an issue](https://github.com/ember-cli/eslint-plugin-ember/issues/202) at the `ember-cli/eslint-plugin-ember` repo,  but have been told there's not much that can be done in code from their side of things. They suggest removing that rule from running on all files in the mirage folder tree.

I think it would be nice to have that rule running, in case any files in that folder tree do extend from Ember's `CoreObject`, and so was wondering if it would be possible to upgrade `Factory` to extend from that so the fix can be applied without having to disable that linting.

Thanks!

**Comment by samselikoff** [comment]
Interesting problem.

The thing is that Mirage was always designed to be decoupled from Ember, so none of the objects used under `/mirage` depend on any Ember classes. There's even been some folks using Mirage outside of Ember in the last 6 months, so I wouldn't want to introduce Ember as a new dependency of Mirage.

Seems like the linter should definitely be able to disambiguate between Ember objects and other objects that just happen to call `.extend()`?

**Comment by rwjblue** [comment]
I do not think that we should change mirage to use `Ember.Object`.  If anything it should be migrated to more clearly use native classes (which would also avoid the linting errors, as these sorts of mistakes are not possible in native ES classes).

**Comment by samselikoff** [comment]
Agreed - thanks for chiming in. Once class props land in stage 3 I would be comfortable moving to ES classes.

**Comment by rwjblue** [comment]
@samselikoff - Based on https://github.com/tc39/proposals it seems that "Class Public Instance Fields & Private Instance Fields" are already Stage 3. 🙀
